enhancements/update/upgrade-without-registry.md

```diff
@@ -0,0 +1,989 @@
+will be shutdown.
+
+The result will be a directory containing approximately 180 images and requires
+16 GiB of space. The command will then create a `upgrade-4.13.7-x86_64.tar` tar
+file containing that directory and a `metadata.json` file.
+
+The `metadata.json` file will contain additional information, in particular the
+architecture, the version, the size and the list of images:
+
+```json
+{
+  "version": "4.13.7",
+  "arch": "x86_64",
+  "size": "16 GiB",
+  "release": "quay.io/openshift-release-dev/ocp-release@sha256:...",
+  "images": [
+    "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+    "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+    ...
+  ]
+}
+```
+
+The `metadata.json` file will always be the first entry of the tar file, to
+simplify operations that need the metadata but not the rest of the contents of
+the tar file.
+
+The command will also write a `upgrade-4.13.7-x86_64.sha256` file containing a
+digest of the complete tar file. This digest is intended to protect the
+integrity of the bundle. It can be checked with tools like `sha256sum`. It will
+also be optionally used by the administrator of the cluster to ensure that the
+right bundle is used in the right cluster. To do so the administrator of the
+cluster will write it to the `desiredUpdate.bundle.digest` field of the
+`ClusterVersion` object. The cluster version operator will reject the bundle if
+it doesn't match this digest.
+
+The engineer will write the tar file to some kind of media and hand it over to
+the technicians, together with the documentation explaining how to use it.
+
+#### Documentation to use the bundle
+
+This documentation shouldn't assume previous OpenShift knowledge, and should be
+basic instructions to plug the USB stick containing the bundle and then check if
+upgrade the upgrade succeeded or failed.
+
+### Risks and Mitigations
+
+The proposed solution will require space to store the release bundle and all the
+release images in all the nodes of the cluster, approximately 48 GiB in the
+worst case. To mitigate that risks the components that will consume disk space
+will check in advance if the required space is available.
+
+### Drawbacks
+
+This approach requires non trivial changes to the cluster version operator and,
+to a lesser degree, to the machine config operator.
+
+## Design Details
+
+### Open Questions
+
+None.
+
+### Test Plan
+
+We should have at least tests that verifies that the upgrade can be performed in
+a fully disconnected environment, both for a single node cluster and a cluster
+with multiple nodes. These tests should gate the OCP release.
+
+It is desirable to have another test that scans the OCP components looking for
+use of the `Always` pull policy. This should probably run for each pull request
+of each OCP component, and prevent merging if it detects that the offending pull
+policy is used. We should consider adding admission in CI for this.
```

Comment:
IIRC, the use case for `Always` is that it forces a check for whether the image (even if present) is accessible by the pod's service account.  Which is important if any secret/proprietary images are possibly present on the node.

But we don't have that problem for images we ship.

(Hmm, offhand I think to implement this check in CI we'd need to gather all the images in the payload, and then look for pods referencing them and look at their pull policy; and this probably generalizes to OLM too)

Reply:
I'd say it would reduce the probability of errors, which is probably good enough. It won't probably catch all cases, as some operators may have conditional code to select the pull policy. That is where scanning the source code helps: it may detect other potential errors, thus reducing the probability even more.